sdks/java/io/rabbitmq/src/test/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqIOTest.java

```diff
@@ -0,0 +1,269 @@
+
+  private static int port;
+  @ClassRule public static TemporaryFolder temporaryFolder = new TemporaryFolder();
+
+  @Rule public transient TestPipeline p = TestPipeline.create();
+
+  private static transient Broker broker;
+
+  @BeforeClass
+  public static void startBroker() throws Exception {
+    try (ServerSocket serverSocket = new ServerSocket(0)) {
+      port = serverSocket.getLocalPort();
+    }
+
+    System.setProperty("derby.stream.error.field", "MyApp.DEV_NULL");
+    broker = new Broker();
+    BrokerOptions options = new BrokerOptions();
+    options.setConfigProperty(BrokerOptions.QPID_AMQP_PORT, String.valueOf(port));
+    options.setConfigProperty(BrokerOptions.QPID_WORK_DIR, temporaryFolder.newFolder().toString());
+    options.setConfigProperty(BrokerOptions.QPID_HOME_DIR, "src/test/qpid");
+    broker.startup(options);
+  }
+
+  @AfterClass
+  public static void stopBroker() {
+    broker.shutdown();
+  }
+
+  @Test
+  public void testReadQueue() throws Exception {
+    final int maxNumRecords = 10;
+    PCollection<RabbitMqMessage> raw =
+        p.apply(
+            RabbitMqIO.read()
+                .withUri("amqp://guest:guest@localhost:" + port)
+                .withQueue("READ")
+                .withMaxNumRecords(maxNumRecords));
+    PCollection<byte[]> output = raw.apply(ParDo.of(new ConverterFn()));
+
+    List<byte[]> records = generateRecords(maxNumRecords);
+    PAssert.that(output).containsInAnyOrder(records);
+
+    ConnectionFactory connectionFactory = new ConnectionFactory();
+    connectionFactory.setUri("amqp://guest:guest@localhost:" + port);
+    Connection connection = connectionFactory.newConnection();
+    Channel channel = connection.createChannel();
+    channel.queueDeclare("READ", false, false, false, null);
+    for (byte[] record : records) {
+      channel.basicPublish("", "READ", null, record);
+    }
+
+    p.run();
+
+    channel.close();
+    connection.close();
+  }
+
+  @Test(timeout = 60 * 1000)
+  public void testReadExchange() throws Exception {
+    final int maxNumRecords = 10;
+    PCollection<RabbitMqMessage> raw =
+        p.apply(
+            RabbitMqIO.read()
+                .withUri("amqp://guest:guest@localhost:" + port)
+                .withExchange("READEXCHANGE", "fanout", "test")
+                .withMaxNumRecords(maxNumRecords));
+    PCollection<byte[]> output = raw.apply(ParDo.of(new ConverterFn()));
+
+    List<byte[]> records = generateRecords(maxNumRecords);
+    PAssert.that(output).containsInAnyOrder(records);
+
+    ConnectionFactory connectionFactory = new ConnectionFactory();
+    connectionFactory.setUri("amqp://guest:guest@localhost:" + port);
+    Connection connection = connectionFactory.newConnection();
+    final Channel channel = connection.createChannel();
+    channel.exchangeDeclare("READEXCHANGE", "fanout");
+    Thread publisher =
+        new Thread(
+            () -> {
+              try {
+                Thread.sleep(5000);
```

Comment:
Why?

Reply:
Maybe a simple while with timeout to check the Qpid status to avoid flaky tests in case 5 sec is not enough for the QPid init time on an overloaded jenkins server.